### DISCUSSION TITLE
why does mem_alloc take very long time sometimes, and it's a random behavior

### DISCUSSION THREAD

**Discussion by guojinrong-nn** [inducer/pycuda#363] [Troubleshooting] [answered]
In my for loop, most calls to mem_alloc only take less than 1ms, but several calls take longer than 200ms, what happens?  How can I fix or avoid it?

**Comment by inducer** [comment, marked as answer]
Memory allocation is shockingly slow in CUDA at times. As a user, I am not aware of any way to fix that, however PyCUDA does offer a way to circumvent it, through the use of memory pools:

https://documen.tician.de/pycuda/util.html#device-based-memory-pool

These work by simply not returning freed memory to the system, but instead retaining the memory and reusing allocations when requests for similarly-sized memory occur later.
